### PR TITLE
Improved user location on Android

### PR DESCRIPTION
**Pull request by funkyfourier**
### Does any other open PR do the same thing?

Not that I can see.

### What issue is this PR fixing?

https://github.com/react-native-community/react-native-maps/issues/2923

Also, it brings some new possibilities for controlling power consumption and accuracy of user location in maps, specifically `userLocationPriority`, `userLocationUpdateInterval` and `userLocationFastestInterval`. This is achieved by setting a custom LocationSource on the GoogleMap object, which exposes these properties and does not cause issue #2923. 

### How did you test this PR?

This PR applies only to Android. I have described in the issue mention above how to recreate the issue, and this PR solves it. It has been tested on a OnePlus 6 and a Sony XZ1.

### CODE REVIEW

File: lib/android/build.gradle [hunk 1]
```diff
@@ -21,6 +21,7 @@ dependencies {
     exclude group: 'com.android.support'
   }
   implementation "com.android.support:appcompat-v7:${safeExtGet('supportLibVersion', '28.0.0')}"
+  implementation "com.google.android.gms:play-services-location:16.0.0"
```
Comment (felipecsl) [inline lib/android/build.gradle]: I guess this should be using the same `${safeExtGet('playServicesVersion', '16.1.0')` mechanism?

File: lib/android/src/main/java/com/airbnb/android/react/maps/AirMapView.java [hunk 2]
```diff
@@ -107,6 +107,7 @@
   private boolean destroyed = false;
   private final ThemedReactContext context;
   private final EventDispatcher eventDispatcher;
+  private FusedLocationSource fusedLocationSource;
```
Comment (felipecsl) [inline lib/android/src/main/java/com/airbnb/android/react/maps/AirMapView.java]: make this `final` too

File: lib/android/src/main/java/com/airbnb/android/react/maps/FusedLocationSource.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import com.google.android.gms.location.LocationResult;
+import com.google.android.gms.location.LocationServices;
+import com.google.android.gms.maps.LocationSource;
+import com.google.android.gms.tasks.OnSuccessListener;
+
+public class FusedLocationSource implements LocationSource {
+
+    private FusedLocationProviderClient fusedLocationClientProviderClient;
+    private LocationRequest locationRequest;
+    private LocationCallback locationCallback;
```
Comment (felipecsl) [inline lib/android/src/main/java/com/airbnb/android/react/maps/FusedLocationSource.java]: final `fusedLocationProviderClient` and `locationRequest `

File: lib/android/build.gradle (hunk 1 above)
Comment (funkyfourier) [reply lib/android/build.gradle]: Actually I discovered that play-services-location does not exactly follow the same versions as the rest of Play Services.

For example, play-services-location does not have a version 16.1.0:
https://mvnrepository.com/artifact/com.google.android.gms/play-services-maps

But for example play-services-maps does:
https://mvnrepository.com/artifact/com.google.android.gms/play-services-maps

Or am I missing something? If not, there are two ways to do it:

1) Use a constant hard coded version like I did.
2) Expose another Gradle variable like "locationServicesVersion" or something like that.

Comment (felipecsl) [reply lib/android/build.gradle]: You're right. I looked directly in the [Google maven repo](https://dl.google.com/dl/android/maven2/index.html) too and it doesnt seem to have version 16.1.0 for `play-services-location`. That's confusing!
Thanks for explaining, this seems reasonable then!

File: lib/android/src/main/java/com/airbnb/android/react/maps/FusedLocationSource.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+import com.google.android.gms.location.FusedLocationProviderClient;
+import com.google.android.gms.location.LocationCallback;
+import com.google.android.gms.location.LocationRequest;
+import com.google.android.gms.location.LocationResult;
+import com.google.android.gms.location.LocationServices;
+import com.google.android.gms.maps.LocationSource;
+import com.google.android.gms.tasks.OnSuccessListener;
+
+public class FusedLocationSource implements LocationSource {
```
Comment (felipecsl) [inline lib/android/src/main/java/com/airbnb/android/react/maps/FusedLocationSource.java]: Just for the record:
This file is indented with 4 spaces while the rest of the project is indented with 2 spaces.
It's non blocking, we should create a separate pull request for automatic style checking/formatting.

**Comment by christopherdro** [comment]
@funkyfourier Could you resolve the conflicts when you get a chance?  Here are the changes that cause the conflict. https://github.com/react-native-community/react-native-maps/commit/801614efcdca00fa2dbad89fca6af1607b4184e1#diff-ac45350dc94234f6d3c8b856aacbc395

**Comment by funkyfourier** [comment]
Was the fixes ok? When will this be merged?

Thank you.

**Comment by christopherdro** [comment]
@funkyfourier Any thoughts on the issue linked to this PR?

**Comment by funkyfourier** [comment]
I'll have a look tomorrow.

**Pull request merged**